### PR TITLE
feat(webhook): add frm webhook support

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR introduces FRM webhooks . Frm webhooks are consumed when payments are held due to manual review pending from FRM processors. When merchant/ Frm provides decision , webhook is triggered . Based on decision we take action whether to approve or decline the payment.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Upon merchant decision on manual review, FRM send us webhooks. Based on FRM webhooks decision we should approve/decline the transaction. 

## How did you test it?

Test cases 
1 . Test signifyd webhooks.(process both incoming and outgoing webhooks)
2. Test any payment processor webhook. 

Screenshot 2024-05-22 at 12 00 32 PM
Screenshot 2024-05-22 at 12 00 47 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/webhooks.rs [hunk 1]
```diff
@@ -677,6 +677,122 @@ pub async fn mandates_incoming_webhook_flow(
...
+                    _,
+                >(
+                    state.clone(),
+                    req_state,
+                    merchant_account.clone(),
+                    key_store.clone(),
+                    payments::PaymentApprove,
+                    api::PaymentsCaptureRequest {
+                        payment_id: payment_attempt.payment_id,
+                        amount_to_capture: Some(payment_attempt.amount),
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/webhooks.rs:715]: Please send `payment_attempt.amount_to_capture` here

File: crates/router/src/core/webhooks.rs [hunk 2]
```diff
@@ -677,6 +677,122 @@ pub async fn mandates_incoming_webhook_flow(
...
+                        cancellation_reason: Some("Rejected by merchant".to_string()),
+                        ..Default::default()
+                    },
+                    services::api::AuthFlow::Merchant,
+                    payments::CallConnectorAction::Trigger,
+                    None,
+                    HeaderPayload::default(),
+                ))
+                .await?;
+                match payments_response {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/webhooks.rs:767]: Can remove this redundant logic and get the payments_response using the event type

File: crates/router/src/core/webhooks.rs [hunk 3]
```diff
@@ -677,6 +677,122 @@ pub async fn mandates_incoming_webhook_flow(
...
+                    services::ApplicationResponse::JsonWithHeaders((payments_response, _)) => {
+                        let payment_id = payments_response
+                            .payment_id
+                            .clone()
+                            .get_required_value("payment_id")
+                            .change_context(errors::ApiErrorResponse::WebhookProcessingFailure)
+                            .attach_printable("payment id not received from payments core")?;
+                        let status = payments_response.status;
+                        let _event_type: Option<enums::EventType> =
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/webhooks.rs:776]: Please trigger an outgoing webhook here

File: crates/router/src/core/webhooks.rs [hunk 4]
```diff
@@ -677,6 +677,122 @@ pub async fn mandates_incoming_webhook_flow(
...
+#[allow(clippy::too_many_arguments)]
+#[instrument(skip_all)]
+pub(crate) async fn frm_incoming_webhook_flow(
+    state: AppState,
+    req_state: ReqState,
+    merchant_account: domain::MerchantAccount,
+    key_store: domain::MerchantKeyStore,
+    source_verified: bool,
+    event_type: webhooks::IncomingWebhookEvent,
+    _request_details: &api::IncomingWebhookRequestDetails<'_>,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/webhooks.rs:689]: Please remove unused fields here

File: crates/router/src/core/webhooks.rs (hunk 3 above)
Comment (srujanchikke) [reply crates/router/src/core/webhooks.rs]: added

File: crates/router/src/core/webhooks.rs [hunk 5]
```diff
@@ -1828,6 +1940,18 @@ pub async fn webhooks_core<W: types::OutgoingWebhookType>(
...
+                req_state,
+                merchant_account,
+                key_store,
+                source_verified,
+                event_type,
+                object_ref_id,
+                business_profile,
+            ))
+            .await
+            .attach_printable("Incoming webhook flow for external authentication failed")?,
```
Comment (Sakilmostak) [inline crates/router/src/core/webhooks.rs:1954]:
```suggestion
            .attach_printable("Incoming webhook flow for fraud check failed")?,
```

**Review by Sakilmostak — CHANGES_REQUESTED** [review]

**Review by kashif-m — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Pull request merged**